pkg/controllers/provisioning/scheduling/nodeset.go

```diff
@@ -76,3 +102,583 @@ func (s *NodeSet) getDaemons(ctx context.Context, constraints *v1alpha5.Constrai
 func (s *NodeSet) Add(node *Node) {
 	s.nodes = append(s.nodes, node)
 }
+
+func (s *NodeSet) SchedulePod(ctx context.Context, pod *v1.Pod) error {
+	// copy the pod as this method modifies the pod's node selectors to force it to meet topology constraints
+	pod = pod.DeepCopy()
+
+	var possibleNodes []*Node
+	for _, node := range s.nodes {
+		if err := node.Compatible(ctx, pod); err == nil {
+			possibleNodes = append(possibleNodes, node)
+		}
+	}
+
+	tightened := s.constraints.Requirements.Add(v1alpha5.NewPodRequirements(pod).Requirements...)
+	var err error
+	possibleNodes, err = s.filterByTopologies(ctx, tightened, pod, possibleNodes)
+	if err != nil {
+		return err
+	}
+
+	if len(possibleNodes) > 0 {
+		sort.Slice(possibleNodes, byPreferredNode(possibleNodes))
+		if err := possibleNodes[0].Add(ctx, pod); err != nil {
+			return fmt.Errorf("adding pod to node, %w", err)
+		}
+	} else {
+		n := NewNode(s.kubeClient, s.constraints, s.daemonResources, s.instanceTypes)
+		if err := n.Add(ctx, pod); err != nil {
+			return fmt.Errorf("adding pod to node, %w", err)
+		}
+		s.nodes = append(s.nodes, n)
+	}
+
+	if err := s.recordTopologyDecisions(pod); err != nil {
+		s.logger.Errorf("recording topology decision, %s", err)
```

Comment:
Why don't we use the logger from context?

Reply:
Removing the stored logger, originally I didn't need context for anything but logging in `Schedule`, but now that it's used for more I'll just use it.